tests/run/carray_slicing.pyx

```diff
@@ -275,3 +275,194 @@ def struct_ptr_iter():
+    for i in range(10):
+        a[i] = i**2
+    return a[8:]
+
+
+def test_c_to_py_coercing_array_in_var():
+    """
+    >>> test_c_to_py_coercing_array_in_var()
+    [64, 81]
+    """
+    cdef int i
+    cdef int a[10]
+    b = 5555
+    for i in range(10):
+        a[i] = i**2
+    b = a[8:]
+    return b
+
+cpdef test_c_to_py_coercing_array_in_var_2():
+    """
+    >>> test_c_to_py_coercing_array_in_var_2()
+    [64, 81]
+    """
+    cdef int i
+    cdef int a[10]
+    b = 5555
+    for i in range(10):
+        a[i] = i**2
+    b = a[8:]
+    return b
+
+
+cpdef tuple test_coercing_c_slice_to_tuple():
+    """
+    >>> test_coercing_c_slice_to_tuple()
+    (64, 81)
+    """
+    cdef int i
+    cdef int a[10]
+    for i in range(10):
+        a[i] = i**2
+    return a[8:]
+
+
+cpdef set test_coercing_c_slice_to_set():
+    """
+    >>> b = test_coercing_c_slice_to_set()
+    >>> isinstance(b, set)
+    True
+    >>> b = list(b)
+    >>> b.sort()
+    >>> b
+    [64, 81]
+    """
+    cdef int i
+    cdef int a[10]
+    for i in range(10):
+        a[i] = i**2
+    return a[8:]
+
+
+cpdef frozenset test_coercing_c_slice_to_frozenset():
+    """
+    >>> b = test_coercing_c_slice_to_frozenset()
+    >>> isinstance(b, frozenset)
+    True
+    >>> b = list(b)
+    >>> b.sort()
+    >>> b
+    [64, 81]
+    """
+
+    cdef int i
+    cdef int a[10]
+    for i in range(10):
+        a[i] = i**2
+    return a[8:]
+
+
+def test_coercing_c_slice_by_casting():
+    """
+    >>> test_coercing_c_slice_by_casting()
+    [25, 36, 49, 64, 81]
+    (25, 36, 49, 64, 81)
+    [25, 36, 49, 64, 81]
+    [25, 36, 49, 64, 81]
+    """
+    cdef int i
+    cdef int a[10]
+    for i in range(10):
+        a[i] = i**2
+
+    print(<list> a[5:])
+    print(<tuple> a[5:])
+    b = <set> a[5:]
+    assert isinstance(b, set)
+    print(sorted(b))
+    b = <frozenset> a[5:]
+    assert isinstance(b, frozenset)
+    print(sorted(b))
+
+
+@cython.wraparound(True)
+def test_coercing_c_slice_out_of_range():
+    """
+    >>> test_coercing_c_slice_out_of_range()
+    Traceback (most recent call last):
+        ...
+    IndexError: Some of slice indexes is out of range
+    """
+    cdef int i
+    cdef int a[10]
+    b = 0
+    for i in range(10):
+        a[i] = i**2
+    return a[11:]
+
+@cython.wraparound(True)
+def test_coercing_c_slice_out_of_range_2():
+    """
+    >>> test_coercing_c_slice_out_of_range_2()
+    Traceback (most recent call last):
+        ...
+    IndexError: Some of slice indexes is out of range
+    """
+    cdef int i
+    cdef int a[10]
+    b = 0
+    for i in range(10):
+        a[i] = i**2
+    return a[:11]
+
+@cython.wraparound(True)
+def test_coercing_c_slice_wrong_range():
+    """
+    >>> test_coercing_c_slice_wrong_range()
+    Traceback (most recent call last):
+        ...
+    IndexError: Wrong range of slice
+    """
+    cdef int i
+    cdef int a[10]
+    b = 0
+    for i in range(10):
+        a[i] = i**2
+    return a[3:2]
```

Comment:
By analogy with Python types like list this should probably just return an empty list.